File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;
import java.util.stream.Collectors;

class Solution {
    public int solution(int[] numbers, int target) {
        int answer = 0;
        List<Integer> covertedNumbers = Arrays.stream(numbers).boxed().collect(Collectors.toList());
        answer = dfs(covertedNumbers,0,0,target);
        return answer;
    }
    
    public int dfs(List<Integer> numbers, int result, int count, int target) {
        if (count == numbers.size()) {
            if (result == target) {
                return 1;
            }
            return 0;
        }
        return dfs(numbers, result + numbers.get(count), count + 1, target) + 
            dfs(numbers, result - numbers.get(count), count + 1, target);
    }
    
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def solution(numbers, target):
    global results
    results = []
    dfs(numbers, 0, 0)
    return results.count(target)

def dfs(numbers, result, depth):
    if (depth >= len(numbers)):
        results.append(result)
        return

    dfs(numbers, result + numbers[depth], depth + 1)
    dfs(numbers, result - numbers[depth], depth + 1)